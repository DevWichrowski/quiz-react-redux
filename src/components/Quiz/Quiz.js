import React, {useEffect, useState} from 'react';
import "./Quiz.scss";
import {quizQuestions} from "../../core/questions";
import {connect} from "react-redux";
import {addPoints, nextQuestion, resetQuiz} from "../../store/actions/quiz.actions";
import ProgressBar from "../ProgressBar/ProgressBar";
import Button from "../shared/Button/Button";

const Quiz = ({questionNumber, submit, resetQuiz, nextQuestion, ...props}) => {
    const [variantChosen, setVariantChosen] = useState(false);
    const [question, setQuestion] = useState(null);
    const quizLength = quizQuestions.length;

    useEffect(() => {
        setQuestion(quizQuestions.find(question => question.id === questionNumber));
        setVariantChosen(false);
    }, [questionNumber]);

    const selectAnswer = e => {
        if (variantChosen) {
            return;
        }

        setVariantChosen(e.variant);

        if (question.correctAnswer === e.variant) {
            submit()
        }
    };

    const reset = () => {
        resetQuiz();
        setVariantChosen(false);
    };

    return (
        <div className="quiz-container">
            {question != null && <ProgressBar step={question?.id}/>}
            {console.log(variantChosen)}
            <div className="quiz">
                <div className="quiz__question">{question?.content}</div>
                <div className="quiz__answers">
                    {question?.answers.map((answer, index) => <Button
                        key={index} value={answer.variant}
                        onClick={e => selectAnswer(answer)} className={`variant ${variantChosen ?
                        question.correctAnswer === answer.variant ?
                            'correct' : 'error' : null}`}>{answer.label}</Button>)}
                </div>
                <div className="question-footer">
                    <Button onClick={reset}>Reset</Button>
                    {variantChosen && questionNumber !== quizLength ?
                        (
                            <Button onClick={nextQuestion}
                                    className="question-footer__next"><span>Next question</span></Button>) :
                        variantChosen && questionNumber === quizLength ?
                            (<Button onClick={nextQuestion} className="question-footer__finish">Finish</Button>) :
                            null}
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = state => ({
    questionNumber: state.quiz.questionNumber
});

const mapDispatchToProps = dispatch => ({
    submit: () => dispatch(addPoints()),
    resetQuiz: () => dispatch(resetQuiz()),
    nextQuestion: () => dispatch(nextQuestion())
});

export default connect(mapStateToProps, mapDispatchToProps)(Quiz)