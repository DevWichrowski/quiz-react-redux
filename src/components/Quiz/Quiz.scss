@import "../../styles/colors";
@import "../../styles/mixins";

.quiz-container {
  padding: 10px 0;
  height: 100%;
  position: relative;

  .quiz {
    padding: 10px 0;
    height: 80%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;

    &__question {
      margin-bottom: 50px;
      font-weight: bold;
      min-height: 40px;
      padding: 0 50px;
    }

    &__answers {
      width: 80%;
      display: flex;
      flex-direction: column;

      .variant {
        width: 100%;
        margin-bottom: 20px;
        min-height: 35px;
      }

      .correct {
        background-color: $green;
      }

      .error {
        background-color: #a80000;
      }

    }

    .question-footer {
      display: flex;
      justify-content: space-between;
      position: absolute;
      padding: 10px 20px;
      bottom: 0;
      width: 100%;
      background-color: $gray;
      border-bottom-left-radius: 10px;
      border-bottom-right-radius: 10px;

      button {
        font-weight: 800;

        &:hover {

        }
      }

      &__next {
        background-color: $blue;
        color: $gray;
        margin-left: 10px;
        transition: all 0.5s;
      }

      &__next span {
        display: inline-block;
        position: relative;
        transition: 0.5s;
      }

      &__next span:after {
        content: '\00bb';
        position: absolute;
        opacity: 0;
        top: 0;
        right: -20px;
        transition: 0.5s;
      }

      &__next:hover span {
        padding-right: 15px;
      }

      &__next:hover span:after {
        opacity: 1;
        right: 0;
      }

      &__finish {

      }
    }
  }
}
