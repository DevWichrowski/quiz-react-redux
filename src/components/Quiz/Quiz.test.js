import Quiz from "./Quiz";
import React from "react";
import {testConnectedComponentRender} from "../../core/helpers/tests/test-connected-component-render";

describe('Quiz', () => {
    testConnectedComponentRender(Quiz)
});
