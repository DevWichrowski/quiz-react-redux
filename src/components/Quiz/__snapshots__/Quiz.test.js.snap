// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz should render 1`] = `
<ContextProvider
  value={null}
>
  <Quiz
    nextQuestion={[Function]}
    questionNumber={0}
    resetQuiz={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
    submit={[Function]}
  />
</ContextProvider>
`;
